crates/re_renderer/src/allocator/data_texture_source.rs

```diff
@@ -0,0 +1,317 @@
+        // If we make buffers always a multiple of this width, we can do all buffer copies in a single copy!
+        //
+        // But wait! If we're using this as the minimum buffer size, isn't that too big?
+        // 16384 * float4 (worst case) = 256KiB.
+        // Keep in mind that weaker hardware will have 8192 max width.
+        // Also note, that many of our textures use 4 & 8 byte formats.
+        // So while this is still a considerable amount of memory when used for very small data textures
+        // it's not as bad as it seems.
+        // Given how much it simplifies to keep buffers a multiple of the texture width,
+        // this seems to be a reasonable trade-off.
+        (self.ctx.device.limits().max_texture_dimension_2d as usize).min(16384)
+    }
+
+    /// The number of elements written so far.
+    #[inline]
+    pub fn num_written(&self) -> usize {
+        self.buffers.iter().map(|b| b.num_written()).sum()
+    }
+
+    /// Reserves space for at least `num_elements` elements.
+    ///
+    /// Creating new buffers is a relatively expensive operation, so it's best to
+    /// reserve gratuitously!
+    /// Ideally, you know exactly how many elements you're going to write and reserve
+    /// accordingly at the start.
+    pub fn reserve(&mut self, num_elements: usize) -> Result<(), CpuWriteGpuReadError> {
+        let remaining_capacity: usize = self.buffers.iter().map(|b| b.remaining_capacity()).sum();
+        if remaining_capacity >= num_elements {
+            return Ok(());
+        }
+
+        // Constraints on the buffer size:
+        // * have at least as many elements as requested
+        // * be at least double the size of the last buffer
+        // * keep it easy to copy to textures by always being a multiple of the maximum row size we use for data textures
+        //      -> this massively simplifies the buffer->texture copy logic!
+        let last_buffer_size = self.buffers.last().map_or(0, |b| b.capacity());
+        let new_buffer_size = (num_elements - remaining_capacity)
+            .max(last_buffer_size * 2)
+            .next_multiple_of(self.max_data_texture_width());
+
+        self.buffers
+            .push(self.ctx.cpu_write_gpu_read_belt.lock().allocate(
+                &self.ctx.device,
+                &self.ctx.gpu_resources.buffers,
+                new_buffer_size,
+            )?);
+
+        // Ensure invariant that the active buffer has some remaining capacity.
+        if self.buffers[self.active_buffer_index].remaining_capacity() == 0 {
+            self.active_buffer_index += 1;
+        }
+
+        Ok(())
+    }
+
+    #[inline]
+    pub fn fill_n(&mut self, element: T, num_elements: usize) -> Result<(), CpuWriteGpuReadError> {
+        self.reserve(num_elements)?;
+
+        let mut num_elements_remaining = num_elements;
+
+        loop {
+            let result =
+                self.buffers[self.active_buffer_index].fill_n(element, num_elements_remaining);
```

Comment:
Same here - "fill" makes me think this is about filling existing memory, not appending new data

```suggestion
                self.buffers[self.active_buffer_index].add_n(element, num_elements_remaining);
```